### PR TITLE
Support tab-separated inputs

### PR DESCRIPTION
**Pull request by snukky**
### Description

This PR adds support for TSV inputs in the trainer, scorer and decoder.
Examples:

	marian -c train.yml --tsv -t file.ende --valid-sets 
    marian-scorer -c score.yml --tsv -t file.ende > scores.txt
	marian-decoder -c ape.yml --tsv < file.srcmt > file.pe

Examples for stdin (`--tsv` is automatically added for `-t stdin`):

    cat file.ende | marian -c train.yml --no-shuffle -t stdin
	cat file.ende | marian-scorer -c score.yml -t stdin > scores.txt

List of changes:
- Added the `--tsv` option for marian and marian-scorer (`-t`), and marian-decoder (`-i`)
- Added the `--tsv-fields` for setting the number of fields in the TSV file
- --tsv-fields is guessed based on the model type, so usually it does not need to be set
- For training with TSV, if vocab(s) does not exist, it is created from appropriate fields in the TSV file (all cases described here: https://github.com/marian-nmt/marian-dev/blob/1d6da8be0cfad2366d866c370d9156402411af41/src/data/corpus_base.cpp#L98 )
- For training with TSV, if no vocab paths are given, a common vocabulary is created (this is different than providing separate training files for source and target)
- For training with TSV, validation sets also have to be in TSV format
- Added support for `-t stdin` in marian and marian-scorer
- For training from stdin, `--no-shuffle` is required, otherwise an appropriate error message is displayed
- For training from stdin, creating vocabs automatically is disabled and an appropriate error message is displayed
- If `-t stdin` is provided, `--tsv` is automatically added
- If `--tsv` is provided and `-t` is empty, `-t stdin` is assumed

Added dependencies: none

### How to test

Run example commands above.
Tested manually and added several regression tests: https://github.com/marian-nmt/marian-regression-tests/pull/45

Tested on Ubuntu 16.04, compiling with:

    cmake .. -DUSE_SENTENCEPIECE=on -DCOMPILE_EXAMPLES=on -DCOMPILE_TESTS=on -DCOMPILE_SERVER=on -DCMAKE_BUILD_TYPE=Release -DUSE_CCACHE=on

The new option potentially does not work with (1) BERT models, (2) factored vocabs, and (3) multi-node training. I don't know how to test these properly.

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

File: src/data/corpus_base.cpp [hunk 1]
```diff
@@ -140,8 +252,7 @@ CorpusBase::CorpusBase(Ptr<Options> options, bool translate)
   }
 
   if(translate) {
-    ABORT_IF(vocabPaths.empty(),
-             "Translating, but vocabularies are not given!");
+    ABORT_IF(vocabPaths.empty(), "Translating, but vocabularies are not given!");
```
Comment (frankseide) [inline src/data/corpus_base.cpp:255]: Consider rewriting the error message. E.g.

Vocabularies must be given for translating.

File: src/data/corpus_base.cpp [hunk 2]
```diff
@@ -298,26 +408,28 @@ void CorpusBase::addWeightsToBatch(Ptr<CorpusBatch> batch,
...
+  // correspoding input type.
+
+  size_t numStreams = tsv_ ? tsvNumFields_ : paths_.size(); // determine number of streams
 
-  addEOS_.resize(paths_.size(), true);
+  addEOS_.resize(numStreams, true);
   // @TODO: think if this should be checked and processed here or in a validation step in config?
   auto inputTypes = options_->get<std::vector<std::string>>("input-types", {}); // empty list by default
 
   // make sure there is an input type for each path
```
Comment (frankseide) [inline src/data/corpus_base.cpp:420]: for each path -> for each stream (here and below)

File: src/data/corpus_base.h [hunk 3]
```diff
@@ -545,6 +545,9 @@ class CorpusBase
   bool maxLengthCrop_{false};
   bool rightLeft_{false};
 
+  bool tsv_{false};          // if the input is a single file with tab-separated values
```
Comment (frankseide) [inline src/data/corpus_base.h:548]: // true if the...

File: src/data/corpus_base.h [hunk 4]
```diff
@@ -545,6 +545,9 @@ class CorpusBase
...
+  size_t tsvNumFields_{0};   // number of fields in the TSV input
```
Comment (frankseide) [inline src/data/corpus_base.h:549]: add:
(only if tsv_)

File: src/training/scheduler.h [hunk 5]
```diff
@@ -18,6 +18,7 @@ class Scheduler : public TrainingObserver {
   std::vector<Ptr<ValidatorBase>> validators_;
 
   bool first_{true};
+  bool endOfStdin_{false};
```
Comment (frankseide) [inline src/training/scheduler.h:21]: This looks fishy. Please add a comment why this variable exists.

File: src/training/scheduler.h [hunk 6]
```diff
@@ -169,8 +169,11 @@ class Scheduler : public TrainingObserver {
 
     // stop if the first validator did not improve for a given number of checks
     size_t stopAfterStalled = options_->get<size_t>("early-stopping");
-    if(stopAfterStalled > 0 && !validators_.empty()
-       && stalled() >= stopAfterStalled)
+    if(stopAfterStalled > 0 && !validators_.empty() && stalled() >= stopAfterStalled)
```
Comment (frankseide) [inline src/training/scheduler.h:172]: This change seems unrelated. Can it be done in a different PR? Or please respond to this comment  to explain what this is about

**Comment by snukky** [comment]
WIP. I will finish and respond to the comments tomorrow.

File: src/common/config.cpp [hunk 7]
```diff
@@ -71,6 +71,20 @@ void Config::initialize(ConfigParser const& cp) {
     }
   }
 
+  // guess --tsv-fields if not set
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config.cpp]: I agree, this is brittle. Alternatives I was thinking about:
- The number of (non-fake) encoders and decoders, but models are being contructed much later than the config and their construction is independent from the corpus.
- The number of (non-zeroed) dim-vocabs loaded from the model special:model.yml, so this will not work for a new training, and LMs are a special case.

I changed the code adding the second approach, but for a fresh training the number of streams is still guessed based on the model type name.

Any ideas how we could determine the number of streams in a cleaner way?

File: src/common/config.cpp [hunk 8]
```diff
@@ -71,6 +71,20 @@ void Config::initialize(ConfigParser const& cp) {
...
+  if(get<bool>("tsv") && get<size_t>("tsv-fields") == 0) {
+    auto modelType = get<std::string>("type");
+
+    size_t tsvFields = 1;
+    if(modelType.find("multi-", 0) != std::string::npos) // is a dual-source model
+      tsvFields += 1;
+    if(mode == cli::mode::training || mode == cli::mode::scoring)
+      if(modelType.rfind("lm", 0) != 0) // unless it is a language model
+        tsvFields += 1;
+
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config.cpp:100]: As you have noticed, alignment and weights are not supported in TSV files.

File: src/common/config_parser.cpp [hunk 9]
```diff
@@ -861,6 +875,21 @@ Ptr<Options> ConfigParser::parseOptions(int argc, char** argv, bool doValidate){
...
 
+  // Option shortcuts for input from STDIN for trainer and scorer
+  if(mode_ == cli::mode::training || mode_ == cli::mode::scoring) {
+    auto trainSets = get<std::vector<std::string>>("train-sets");
+    YAML::Node config;
+    // Assume the input will come from STDIN if --tsv is set but no --train-sets are given
+    if(get<bool>("tsv") && trainSets.empty()) {
+      config["train-sets"].push_back("stdin");
+    // Assume the input is in TSV format if --train-sets is set to "stdin"
+    } else if(trainSets.size() == 1 && trainSets[0] == "stdin") {
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_parser.cpp]: Added. However, I think "-" might not be supported everywhere in the existing code.

File: src/common/config_parser.cpp [hunk 10]
```diff
@@ -861,6 +875,21 @@ Ptr<Options> ConfigParser::parseOptions(int argc, char** argv, bool doValidate){
...
+      config["tsv"] = true;
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_parser.cpp:887]: It works, and I would say it's correct. My idea was that for simplicity and to avoid potential special cases, it might be better to set TSV always if the training is from stdin, regardless of the number of streams.

I also added a regression test for training with `cat file | marian --type lm -t stdin`.

[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_parser.cpp:887]: BTW, at the moment marian does not know about the correct number of streams for different model types and does not put restrictions on the number of streams. The only validation is len(train-set) == len(vocabs). It's possible to train a LM with parallel data, while training a seq2seq model with a single train-set and vocab results in a segmentation fault. This probably deserves an issue on GitHub.

File: src/common/config_validator.cpp [hunk 11]
```diff
@@ -70,9 +70,19 @@ void ConfigValidator::validateOptionsParallelData() const {
...
            "There should be as many vocabularies as training sets");
+
+  // disallow, for example --tsv --train-sets file1.tsv file2.tsv
+  ABORT_IF(get<bool>("tsv") && trainSets.size() != 1,
+      "A single file can be provided with --train-sets (or stdin) for a tab-separated input");
+
+  // disallow, for example --train-sets stdin stdin or --train-sets stdin file.tsv
+  ABORT_IF(trainSets.size() > 1 && std::any_of(trainSets.begin(),
+                                               trainSets.end(),
+                                               [](const std::string& s) { return s == "stdin"; }),
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_validator.cpp]: Done

File: src/common/config_validator.cpp [hunk 12]
```diff
@@ -70,9 +70,19 @@ void ConfigValidator::validateOptionsParallelData() const {
...
 
-  auto vocabs = get<std::vector<std::string>>("vocabs");
-  ABORT_IF(!vocabs.empty() && vocabs.size() != trainSets.size(),
+  auto numVocabs = get<std::vector<std::string>>("vocabs").size();
+  ABORT_IF(!get<bool>("tsv") && numVocabs > 0 && numVocabs != trainSets.size(),
            "There should be as many vocabularies as training sets");
+
+  // disallow, for example --tsv --train-sets file1.tsv file2.tsv
+  ABORT_IF(get<bool>("tsv") && trainSets.size() != 1,
+      "A single file can be provided with --train-sets (or stdin) for a tab-separated input");
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_validator.cpp]: Done

File: src/common/config_validator.cpp [hunk 13]
```diff
@@ -105,10 +115,15 @@ void ConfigValidator::validateOptionsTraining() const {
   ABORT_IF(!modelDir.empty() && !filesystem::isDirectory(modelDir),
            "Model directory does not exist");
 
+  std::string errorMessage = get<bool>("tsv")
+            ? "There should be as many validation sets as training sets. "
+              "If the training set is in the TSV format, validation sets have to also be a TSV file"
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_validator.cpp]: Done

File: src/common/config_validator.cpp [hunk 14]
```diff
@@ -105,10 +115,15 @@ void ConfigValidator::validateOptionsTraining() const {
   ABORT_IF(!modelDir.empty() && !filesystem::isDirectory(modelDir),
            "Model directory does not exist");
 
+  std::string errorMessage = get<bool>("tsv")
+            ? "There should be as many validation sets as training sets. "
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/config_validator.cpp]: Renamed "training/validation sets" to "training/validation files" wherever appropriate in this file.

File: src/common/file_utils.cpp [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+void cut(const std::string& tsvIn,
+         Ptr<io::TemporaryFile> tsvOut,
+         const std::vector<size_t>& fields,
+         size_t numFields,
+         const std::string& sep /*= "\t"*/) {
+  io::InputFileStream ioIn(tsvIn);
+  std::string line;
+  while(getline(ioIn, line)) {
+    std::vector<std::string> tsvFields(numFields);
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/file_utils.cpp]: Moved outside the loop.

File: src/common/utils.cpp [hunk 16]
```diff
@@ -67,6 +67,24 @@ void split(const std::string& line,
   }
 }
 
+void splitTsv(const std::string& line, std::vector<std::string>& fields, size_t numFields) {
+  fields.clear();
+  if(fields.capacity() != numFields) // make sure the number of fields is always as requested
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp]: Thanks! I changed that to resizing the vector at the end of the function, to guarantee that the output has as many elements as requested. I am not sure if it's actually necessary, but I think it's safer.

Added also a regression test with excessive tabs and simple unit tests.

File: src/common/utils.cpp [hunk 17]
```diff
@@ -67,6 +67,24 @@ void split(const std::string& line,
...
+  for(size_t i = 0; i < numFields; ++i) {
+    pos = line.find('\t', begin);
+    if(pos == std::string::npos) {
+      fields.push_back(line.substr(begin));
+      break;
+    }
+    fields.push_back(line.substr(begin, pos - begin));
+    begin = pos + 1;
+  }
+}
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp:91]: I added a warning if there is an excessive tab-separated field.

Added also a regression test with excessive tabs and simple unit tests.

File: src/data/corpus.cpp [hunk 18]
```diff
@@ -78,13 +78,22 @@ SentenceTuple Corpus::next() {
...
-      if(i > 0 && i == alignFileIdx_) { // @TODO: alignFileIdx == 0 possible?
+      if(i > 0 && i == alignFileIdx_) {
         addAlignmentToSentenceTuple(line, tup);
       } else if(i > 0 && i == weightFileIdx_) {
         addWeightsToSentenceTuple(line, tup);
       } else {
-        preprocessLine(line, i);
-        addWordsToSentenceTuple(line, i, tup);
+        if(tsv_) {  // split TSV input and add each field into the sentence tuple
+          std::vector<std::string> fields(tsvNumFields_);
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus.cpp]: Done

File: src/data/corpus.cpp [hunk 19]
```diff
@@ -78,13 +78,22 @@ SentenceTuple Corpus::next() {
         }
       }
 
-      if(i > 0 && i == alignFileIdx_) { // @TODO: alignFileIdx == 0 possible?
+      if(i > 0 && i == alignFileIdx_) {
         addAlignmentToSentenceTuple(line, tup);
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus.cpp:84]: Correct, support for alignment and data weighting in TSV files is not a part of this PR.

File: src/data/corpus.cpp [hunk 20]
```diff
@@ -143,6 +153,10 @@ void Corpus::restore(Ptr<TrainingState> ts) {
 void Corpus::shuffleData(const std::vector<std::string>& paths) {
   LOG(info, "[data] Shuffling data");
 
+  ABORT_IF(tsv_ && paths[0] == "stdin",
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus.cpp]: Done

File: src/data/corpus_base.cpp [hunk 21]
```diff
@@ -36,9 +37,16 @@ CorpusBase::CorpusBase(const std::vector<std::string>& paths,
...
       maxLength_(options_->get<size_t>("max-length")),
       maxLengthCrop_(options_->get<bool>("max-length-crop")),
-      rightLeft_(options_->get<bool>("right-left")) {
-  ABORT_IF(paths_.size() != vocabs_.size(),
-           "Number of corpus files and vocab files does not agree");
+      rightLeft_(options_->get<bool>("right-left")),
+      tsv_(options_->get<bool>("tsv", false)),
+      tsvNumFields_(options->get<size_t>("tsv-fields", 0)) {
+  if(tsv_) {
+    ABORT_IF(tsvNumFields_ != vocabs_.size(),
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp:44]: Yes, I think so. At the moment, however, marian does not know about the correct number of streams and assumes that there is as many streams as vocabs or train-sets (see one of the previous comments about potential errors that this generates).

File: src/data/corpus_base.cpp [hunk 22]
```diff
@@ -68,69 +78,171 @@ CorpusBase::CorpusBase(Ptr<Options> options, bool translate)
...
+    //  * cat file.tsv | --tsv -t stdin
+    //    As above, an error is shown that creation of a vocabulary from stdin is not supported.
+    //
+    //  There is more cases for multi-encoder models not listed above.
+    //
     if(vocabPaths.empty()) {
+      size_t numStreams = tsv_ ? tsvNumFields_ : paths_.size();
+
+      // Creating a vocabulary from stdin is not supported
+      ABORT_IF(tsv_ && paths_[0] == "stdin",
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp]: Supported everywhere in this file.

File: src/data/corpus_base.cpp [hunk 23]
```diff
@@ -68,69 +78,171 @@ CorpusBase::CorpusBase(Ptr<Options> options, bool translate)
...
+        maxVocabs.resize(numStreams, 0);
 
-      // Helper object to for grouping training data based on vocabulary file name
-      struct PathsAndSize {
-        std::set<std::string> paths; // contains all paths that are used for training the vocabulary
-        size_t size;                 // contains the maximum vocabulary size
+      // Helper object for grouping training data based on vocabulary file name
+      struct VocabDetails {
+        std::set<std::string> paths;    // all paths that are used for training the vocabulary
+        std::vector<size_t> positions;  // all position of the vocabulary in --vocab
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp]: Sorry, fixed.

File: src/data/corpus_base.cpp [hunk 24]
```diff
@@ -68,69 +78,171 @@ CorpusBase::CorpusBase(Ptr<Options> options, bool translate)
...
+        size_t size;                    // the maximum vocabulary size
       };
 
-      // Group training files based on vocabulary path. If the same
-      // vocab path corresponds to different training files, this means
-      // that a single vocab should combine tokens from all files.
-      std::map<std::string, PathsAndSize> groupVocab;
+      // Group training files based on vocabulary path. If the same vocab path corresponds to
+      // different training files, this means that a single vocab should combine tokens from all
+      // files.
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp]: Reverted

File: src/data/corpus_base.cpp (hunk 1 above)
Comment (snukky) [reply src/data/corpus_base.cpp:255]: I will open a new PR with this change as it requires updates of regression tests.

File: src/data/corpus_base.h (hunk 3 above)
Comment (snukky) [reply src/data/corpus_base.h]: Done

File: src/data/corpus_base.cpp (hunk 2 above)
Comment (snukky) [reply src/data/corpus_base.cpp]: Done

File: src/data/corpus_base.h (hunk 4 above)
Comment (snukky) [reply src/data/corpus_base.h]: Done

File: src/training/scheduler.h (hunk 5 above)
Comment (snukky) [reply src/training/scheduler.h]: It helps to properly stop the training when the data stream from stdin is over (otherwise there would be an infinite loop of `corpus->next`) and this works with (1) training for multiple epochs and (2) renewing the training. Simply `if(tsv_ && state_->epoch > 1)` does not work with (1), and relying on the difference between `state_->epochs` and `state_->prevEpochs` did not work with (2).

I tried to put this into a comment in the code.

File: src/training/scheduler.h (hunk 6 above)
Comment (snukky) [reply src/training/scheduler.h]: Reverted. Such kind of changes are usually a mix of the boy scout rule and hitting the autoformat shortkey too frequently. I agree that changes should be kept minimal.

File: src/data/corpus_base.cpp [hunk 25]
```diff
@@ -68,69 +78,171 @@ CorpusBase::CorpusBase(Ptr<Options> options, bool translate)
...
+    //    from stdin is not supported.
+    //
+    //  * cat file.tsv | --tsv -t stdin -v vocab vocab
+    //    If vocab does not exist, an error is shown that creation of a vocabulary from stdin is not
+    //    supported.
+    //
+    //  * cat file.tsv | --tsv -t stdin
+    //    As above, an error is shown that creation of a vocabulary from stdin is not supported.
+    //
+    //  There is more cases for multi-encoder models not listed above.
```
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp:132]: (please see the comment above)

**Comment by snukky** [comment]
Many thanks for the review, @frankseide! I have addressed all comments, and left a few of them unresolved, where I thought you may want to take a look again.

File: src/common/config.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (frankseide) [reply src/common/config.cpp]: Unfortunately not...

File: src/common/utils.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (frankseide) [reply src/common/utils.cpp:91]: Let's make this a hard error

File: src/common/utils.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (frankseide) [reply src/common/utils.cpp]: Somehow I cannot see the changed code in here, but let's make it a hard error. Either the file contains the correct number of TAB-separated fields, or it is malformed, and the training should terminate.

We could consider skipping a line with a warning, but I suggest to make it a hard error for now until someone shouts about it.

File: src/data/corpus_base.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (frankseide) [reply src/data/corpus_base.cpp:44]: Could we add a TODO comment to the source code for this?

File: src/common/utils.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp]: GitHub is not good at presenting changes for a review. It's here: https://github.com/marian-nmt/marian-dev/blob/6da35a2d8dd9f9a2db503d4bb954d72b3a8a9c01/src/common/utils.cpp#L70-L91

I think that a hard error might be too much. I would be annoyed if my job waiting for two days in a slurm queue is aborted because only of a few malformed lines in the corpus. If there is many malformed lines, the logging file will be spammed with warning messages, for example: https://github.com/marian-nmt/marian-regression-tests/blob/325fd45d0a8a0648c39cc26e9757b8bee246d20d/tests/interface/input-tsv/train_extra_tabs.expected

At the moment there is a warning and the line is not skipped, but I now think that it should be.

[the remark this answers is not in the document]
Comment (kpu) [reply src/common/utils.cpp]: My `StringPiece` / `string_ref` tic is triggered by this function declaration.

[the remark this answers is not in the document]
Comment (kpu) [reply src/common/utils.cpp]: Also, I vote for hard error.

[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp]: We have warnings for other failed checks, e.g. a missing sentence weight, missing or excessive word weights. Maybe these should be hard errors too? OK, if I'm not wrong, incorrect word alignment during guided alignment is a segmentation fault, so it's actually a hard error...

[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp]: Replaced the warning with an abort.

File: src/common/utils.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (snukky) [reply src/common/utils.cpp:90]: Added an abort.

File: src/data/corpus_base.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (snukky) [reply src/data/corpus_base.cpp:45]: Added

**Comment by frankseide** [comment]
Roman, could you please confirm that the test suite passes, and only then Merge it?

**Comment by snukky** [comment]
Yes, I am running the tests again. Thanks for the review! I will continue with supporting alignment and weights in TSV inputs.

@emjotde Do you have any comments with the PR? (`@marcin` from Warsaw has to be getting a lot of unrelated mentions because of his username...)

**Comment by emjotde** [comment]
Looking now.

**Comment by emjotde** [comment]
@snukky Great work here. We will extensively test this by using it a lot now :)
